contents/go/Go.md

```diff
@@ -0,0 +1,143 @@
+
+More information on these three mechanisms and their uses can be found on the [Go blog](https://blog.golang.org/defer-panic-and-recover).
+
+### Interfaces
+Although Go has types and methods and allows pseudo-object-oriented style of programming, type hierarchy does not exist in Go. Instead, Go uses interfaces to specify methods that types should implement, grouping types by functionality rather than using notions of inheritance. Types implement interfaces by implementing the methods in the interface, and do not need to explicitly specify which interfaces are implemented.
+
+In the example below, the `Rectangle` type implements the interface `TwoDimensional` by implementing the methods `area()` and `perim()` that are specified in the interface.
+
+```go
+type TwoDimensional interface {
+  area() float64
+  perim() float64
+}
+
+type Rectangle struct {
+  width, height float64
+}
+
+func (r Rectangle) area() {
+  return r.width * r.height
+}
+
+func (r Rectangle) perim() {
+  return r.width * 2 + r.height * 2
+}
+```
+
+One benefit of using a system where interface implementations need not be stated in the source code is that methods can be attached to types that you didn't write. In other words, you can extend a type to implement an interface without access to its source code by simply implementing the interface's method in your own code.
+
+Some resources to get started with Go interfaces include [this blog post](https://medium.com/golangspec/interfaces-in-go-part-i-4ae53a97479c) introducing Go interfaces, code examples on
+[how interfaces (including the empty interface) are used in practice](https://www.calhoun.io/how-do-interfaces-work-in-go/), this [comparison of Go's OOP style with that of other languages](https://flaviocopes.com/golang-is-go-object-oriented/) and [Go's official FAQ on OOP](https://golang.org/doc/faq#Is_Go_an_object-oriented_language).
+
+### Concurrency
+One of Go's special features is its support for concurrency. To this end, Go's standard library comes with two features that allow for easy and maintainable concurrency.
+
+#### Goroutines
+A goroutine is a lightweight thread that executes a function concurrently with its caller. A goroutine is launched by a `go` statement:
+
+```go
+func run() {
+  \\ does something
+}
+
+func main() {
+  go run()
+  executeOtherCommands()
+}
+```
+
+In the above example, using the `go` keyword launches a goroutine, which executes `run()` concurrently with `executeOtherCommands()`.
+
+Goroutines can also be started for anonymous functions:
+```go
+go func(msg string) {
+  fmt.Println(msg)
+}("a message")  // starts a goroutine that prints "a message"
+```
+
+A goroutine is not its own thread; instead, goroutines are dynamically multiplexed onto threads as required to keep them running. This makes them lightweight, so having a large number of goroutines is feasible. In practice, goroutines behave similarly to very cheap threads.
```

Comment:
You can also specify that goroutines start out with a very small stack (on the order of kBs instead of an OS thread in MBs); see [here](https://stackoverflow.com/questions/41906357/why-are-goroutines-much-cheaper-than-threads-in-other-languages)